libbeat/common/file/interval_rotator.go

```diff
@@ -0,0 +1,209 @@
+}
+
+func (r *intervalRotator) initialize() error {
+	r.clock = realClock{}
+
+	switch r.interval {
+	case time.Second:
+		r.fileFormat = "2006-01-02-15-04-05"
+		r.newInterval = newSecond
+	case time.Minute:
+		r.fileFormat = "2006-01-02-15-04"
+		r.newInterval = newMinute
+	case time.Hour:
+		r.fileFormat = "2006-01-02-15"
+		r.newInterval = newHour
+	case 24 * time.Hour: // calendar day
+		r.fileFormat = "2006-01-02"
+		r.newInterval = newDay
+	case 7 * 24 * time.Hour: // calendar week
+		r.fileFormat = ""
+		r.newInterval = newWeek
+		r.weekly = true
+	case 30 * 24 * time.Hour: // calendar month
+		r.fileFormat = "2006-01"
+		r.newInterval = newMonth
+	case 365 * 24 * time.Hour: // calendar year
+		r.fileFormat = "2006"
+		r.newInterval = newYear
+	default:
+		r.arbitrary = true
+		r.fileFormat = "2006-01-02-15-04-05"
+		r.newInterval = func(lastTime time.Time, currentTime time.Time) bool {
+			lastInterval := lastTime.Unix() / (int64(r.interval) / int64(time.Second))
+			currentInterval := currentTime.Unix() / (int64(r.interval) / int64(time.Second))
+			return lastInterval != currentInterval
+		}
+	}
+	return nil
+}
+
+func (r *intervalRotator) LogPrefix(filename string, modTime time.Time) string {
+	var t time.Time
+	if r.lastRotate.IsZero() {
+		t = modTime
+	} else {
+		t = r.lastRotate
+	}
+
+	if r.weekly {
+		y, w := t.ISOWeek()
+		return fmt.Sprintf("%s-%04d-%02d-", filename, y, w)
+	}
+	if r.arbitrary {
+		intervalNumber := t.Unix() / (int64(r.interval) / int64(time.Second))
+		intervalStart := time.Unix(0, intervalNumber*int64(r.interval))
+		return fmt.Sprintf("%s-%s-", filename, intervalStart.Format(r.fileFormat))
+	}
+	return fmt.Sprintf("%s-%s-", filename, t.Format(r.fileFormat))
+}
+
+func (r *intervalRotator) NewInterval() bool {
+	now := r.clock.Now()
+	newInterval := r.newInterval(r.lastRotate, now)
+	return newInterval
+}
+
+func (r *intervalRotator) Rotate() {
+	r.lastRotate = r.clock.Now()
+}
+
+func (r *intervalRotator) SortIntervalLogs(strings []string) {
+	sort.Slice(
+		strings,
+		func(i, j int) bool {
+			return OrderIntervalLogs(strings[i]) < OrderIntervalLogs(strings[j])
+		},
+	)
+}
+
+// OrderIntervalLogs, when given a log filename in the form [prefix]-[formattedDate]-n
+// returns the filename after zero-padding the trailing n so that foo-[date]-2 sorts
+// before foo-[date]-10.
+func OrderIntervalLogs(filename string) string {
+	index, i, err := IntervalLogIndex(filename)
+	if err == nil {
+		return filename[:i] + fmt.Sprintf("%020d", index)
+	}
+
+	return ""
+}
+
+// IntervalLogIndex returns n as int given a log filename in the form [prefix]-[formattedDate]-n
+func IntervalLogIndex(filename string) (uint64, int, error) {
+	i := len(filename) - 1
+	for ; i >= 0; i-- {
+		if '0' > filename[i] || filename[i] > '9' {
+			break
+		}
+	}
+	i++
+
+	s64 := filename[i:]
+	u64, err := strconv.ParseUint(s64, 10, 64)
+	return u64, i, err
+}
+
+func newSecond(lastTime time.Time, currentTime time.Time) bool {
```

Comment:
To be sure that __nothing__ can go wrong here, you could add a sanity check to see if `currentTime` is indeed later in time than `lastTime`. Otherwise the checks only check if the two times are different, not later.

Reply:
Yes, I also had concerns about that. I'm fine with this.